zone-setup/src/bin/zone-setup.rs

```diff
@@ -4,285 +4,185 @@
+use anyhow::{anyhow, Context};
+use clap::{ArgAction, Args, Parser, Subcommand};
 use illumos_utils::ipadm::Ipadm;
 use illumos_utils::route::{Gateway, Route};
 use illumos_utils::svcadm::Svcadm;
 use omicron_common::cmd::fatal;
 use omicron_common::cmd::CmdError;
 use slog::{info, Logger};
+use std::fmt::Write as _;
 use std::fs::{metadata, read_to_string, set_permissions, write, OpenOptions};
-use std::io::Write;
+use std::io::Write as _;
 use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
 use std::os::unix::fs::chown;
-use std::path::Path;
+use std::path::PathBuf;
 use uzers::{get_group_by_name, get_user_by_name};
 
-pub const HOSTS_FILE: &str = "/etc/inet/hosts";
-pub const CHRONY_CONFIG_FILE: &str = "/etc/inet/chrony.conf";
-pub const LOGADM_CONFIG_FILE: &str = "/etc/logadm.d/chrony.logadm.conf";
-pub const ROOT: &str = "root";
-pub const SYS: &str = "sys";
-
-pub const COMMON_NW_CMD: &str = "common-networking";
-pub const OPTE_INTERFACE_CMD: &str = "opte-interface";
-pub const CHRONY_SETUP_CMD: &str = "chrony-setup";
-
-fn parse_ip(s: &str) -> anyhow::Result<IpAddr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IP address"))
+#[derive(Debug, Parser)]
+struct ZoneSetup {
+    #[command(subcommand)]
+    command: ZoneSetupCommand,
 }
 
-fn parse_ipv4(s: &str) -> anyhow::Result<Ipv4Addr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IPv4 address"))
+#[derive(Debug, Subcommand)]
+enum ZoneSetupCommand {
+    /// Sets up common networking configuration across all zones
+    CommonNetworking(CommonNetworkingArgs),
+    /// Sets up OPTE interface
+    OpteInterface(OpteInterfaceArgs),
+    /// Sets up Chrony configuration for NTP zone
+    ChronySetup(ChronySetupArgs),
 }
 
-fn parse_ipv6(s: &str) -> anyhow::Result<Ipv6Addr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IPv6 address"))
+// Some of our `String` cli arguments may be set to the literal "unknown" if
+// sled-agent hasn't populated our SMF config correctly. Reject this at argument
+// parsing time.
+fn parse_string_rejecting_unknown(s: &str) -> anyhow::Result<String> {
+    if s != "unknown" {
+        Ok(s.to_string())
+    } else {
+        Err(anyhow!("missing input value"))
+    }
 }
 
-fn parse_datalink(s: &str) -> anyhow::Result<String> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing data link"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid data link"))
+#[derive(Debug, Args)]
+struct CommonNetworkingArgs {
+    #[arg(short, long, value_parser = parse_string_rejecting_unknown)]
+    datalink: String,
+    #[arg(short, long)]
+    gateway: Ipv6Addr,
+    #[arg(short, long)]
+    static_addr: Ipv6Addr,
```

Comment:
I'm pretty sure some of these flags (and in the other commands as well) are required.

Reply:
All the flags are still required, except for `--allow` for chrony-setup (because its type is `Option<...>`).

Reply:
Ahhhh gotcha gotcha, I think I was expecting a `required = true`, but if just setting without `Option<...>` has the same result, then it's all good